src/_pytest/cacheprovider.py

```diff
@@ -178,20 +183,13 @@ def set(self, key: str, value: object) -> None:
         """
         path = self._getvaluepath(key)
         try:
-            if path.parent.is_dir():
-                cache_dir_exists_already = True
-            else:
-                cache_dir_exists_already = self._cachedir.exists()
-                path.parent.mkdir(exist_ok=True, parents=True)
+            self._mkdir(path.parent)
```

Comment:
While it doesn't matter much, it doesn't feel right to me to repeatedly do the `_ensure_cache_dir_and_supporting_files` work on `set`. 

A better solution I think is to do the cache dir creation atomic, i.e. create the dir in a `TemporaryDirectory` (in same dir as the cache dir so it's in the same filesystem) then "commit" it by renaming to `.pytest_cache`.

Reply:
Hmm that will complicate things a bit if the user is (for some reason) using `cachedir=.` (as in the tests for this).

Reply:
Actually if the user is using an existing directory for this.

Reply:
How does it complicate things? (With the atomic solution the dir is not touched if it already exists)

Reply:
IIUC we always want to add the `CACHEDIR.TAG`, no?

Reply:
It's not really clear what we should do when operating in an existing directory. Presumably we don't want to write `.gitignore` because we may end up ignoring the user's files, but we probably do want to write `CACHEDIR.TAG`.

The previous implementation didn't write anything, if the directory already existed. The current implementation tries to write everything, so long as it's not overwriting. Doing all the work in a temporary directory would amount to the previous behavior. What do we want the behavior to be?

Reply:
IMO the previous behavior was good, if the directory already exists, don't add the gitignore/CACHEDIR/README.

Reply:
Done. Note that the solution you originally suggested (creating the directory elsewhere, then moving it) isn't feasible because there's no atomic move operation for directories. In other words, the best we can do is create all the files *before* writing the cache data, greatly reducing the probability of a half-created cache directory but not eliminating it. That's what I've done now.

Reply:
> IMO the previous behavior was good, if the directory already exists, don't add the gitignore/CACHEDIR/README.

Hmm but isn't this what this PR attempts to fix? The objective is to write the files in case the directory was created by pytest but the files were not written because the session was interrupted suddenly.

If we want the previous behavior, probably best to close the PR at all IMHO.

Reply:
No, the previous behavior was:

- create the directory
- attempt to write the cache data
- create supporting files

the problem this PR solves is that writing the cache data can take a comparatively long time; by writing supporting files immediately after creating the directory, we narrow the opportunity for corruption significantly.

Reply:
Sorry, I was mistaken. Indeed, it seems the race between directory creation and initialization of supporting files is significant enough that even when everything happens in sequence we have a problem. That is unless the problem I observed locally arose from `def mkdir` which prior to this PR did not properly initialize the cache directory at all.

Reply:
Indeed, in the previous version of the code, we did write the supporting files right after creating the directory, and before attempting to write the cache data. 

At first your issue seemed curious to me, because the session would have to be interrupted exactly between `mkdir` and `ensure_supporting_files`, which seems unlikely (but theoretically possible). But ensuring the files are always there seemed OK too.

But if @bluetech thinks the current behavior (only create the files if we create the directory) is OK, then we should probably close this PR instead of introducing these changes (which are OK but do not change the underlying behavior, so I would rather not change anything).

Reply:
> there's no atomic move operation for directories

Are you sure? AFAIU, `os.rename` (or `os.replace`) on a directory is atomic. It just renames the directory entry, doesn't move files etc (as long as it's on the same filesystem).

Reply:
@bluetech I think you're right. The last remaining difficulty is that [`tempfile.TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) requires you to specify whether it should attempt cleanup at the time of its construction. In our case we want clean to be performed if we're interrupted, but not after we've successfully initialized and moved the directory. I've done the closest thing: ignore cleanup errors. PTAL.

Reply:
At long last: this is green.